File: src/main/java/hex/paymentstore/adapter/PaymentController.java
package hex.paymentstore.adapter;

import hex.paymentstore.domain.AddPaymentUseCase;
import hex.paymentstore.domain.PaymentRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.Errors;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import javax.validation.Valid;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.NOT_FOUND;

@RestController
public class PaymentController {
	private final PaymentRepository repository;
	private final AddPaymentUseCase addPaymentUseCase;
	
	public PaymentController(PaymentRepository repository, AddPaymentUseCase addPaymentUseCase) {
		this.repository = repository;
		this.addPaymentUseCase = addPaymentUseCase;
	}
	
	@PostMapping
	public ResponseEntity<?> add(@Valid @RequestBody RestPayment payment, Errors errors) {
		if(errors.hasErrors()) {
			return ResponseEntity.badRequest().body(errors.getAllErrors().stream().map(ObjectError::getDefaultMessage).collect(Collectors.joining()));
		}
		RestPayment body = new RestPayment(addPaymentUseCase.add(payment.toDomain()));
		return new ResponseEntity<>(body, CREATED);
	}
	
	@GetMapping
	public List<RestPayment> getPayments() {
		return repository.getAll().stream().map(RestPayment::new).collect(toList());
	}
	
	@GetMapping("/{id}")
	public RestPayment getPayment(@PathVariable String id) {
		return repository.find(id)
				.map(RestPayment::new)
				.orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Unable to find payment with id=$id"));
	}
	
	@DeleteMapping("/{id}")
	public void delete(@PathVariable String id) {
		repository.remove(id);
	}
}

File: src/main/kotlin/hex/paymentstore/adapter/PaymentController.kt
package hex.paymentstore.adapter

import hex.paymentstore.domain.AddPaymentUseCase
import hex.paymentstore.domain.PaymentRepository
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.http.ResponseEntity
import org.springframework.validation.Errors
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import javax.validation.Valid

@RestController
class PaymentController(
	private val repository: PaymentRepository,
	private val addPaymentUseCase: AddPaymentUseCase
){
	
	@PostMapping
	fun add(@Valid @RequestBody payment: RestPayment, errors: Errors): ResponseEntity<Any> {
		if(errors.hasErrors()) {
			return ResponseEntity.badRequest().body(errors.getAllErrors().map { it.defaultMessage }.toString())
		}
		val body = addPaymentUseCase.add(payment.toDomain()).let(::RestPayment)
		return ResponseEntity(body, CREATED)
	}
	
	@GetMapping
	fun getPayments(): List<RestPayment> = repository.getAll().map(::RestPayment)
	
	
	@GetMapping("/{id}")
	fun getPayment(@PathVariable id: String): RestPayment =
			repository.find(id)
					?.let(::RestPayment)
					?: throw ResponseStatusException(NOT_FOUND, "Unable to find payment with id=$id")
	
	
	@DeleteMapping("/{id}")
	fun delete(@PathVariable id: String) = repository.remove(id)
}